### PR TITLE
Fixed topographic_error() and quantization_error()

### PR DESCRIPTION
**Pull request by wei-zhang-thz**
Problems:
- The previous topographic_error() method is incorrect. bmu_1 and bmu_2 are not the coordinates of the best two matching units.
- The previous topographic_error() and quantization_error() uses explicit for-loops, which is very slow.

Fixes:
- Fixed incorrect implementation of topographic_error() method.
- Changed the topographic_error() and quantization_error() methods with vectorized implementation.

### CODE REVIEW

**Comment by JustGlowing** [comment]
Hi there, thanks for your submission. I'll have a look next week.

**Comment by JustGlowing** [comment]
In the meanwhile, please make sure that the methods you are introducing are unit tested and make sure that pycodestyle doesn't flag any line.

**Comment by wei-zhang-thz** [comment]
Thanks. The original commit only works for numpy.ndarray type input. I fixed this problem and now it passes the unit test.

**Comment by JustGlowing** [comment]
That's amazing!

Can you report there speedup achieved on the quantization error?

Also, please don't disable any of the pylint checks.

**Comment by wei-zhang-thz** [comment]
The speed-up is very significant. On my PC, it's about 50 times faster. When the data is 300-by-50000, original methods take several minutes, while the new methods take seconds to complete.

The reason that I disabled the pylint check is because I encountered some issue similar to this: https://github.com/PyCQA/pylint/issues/2061. Not sure how to resolve it. Do you have suggestions on that?

**Comment by JustGlowing** [comment]
I would just avoid unpacking the two elements in a single line.

**Comment by wei-zhang-thz** [comment]
Please checkout the new version. Thanks!

File: minisom.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from math import sqrt
 
+import numpy as np
```
Comment (JustGlowing) [inline minisom.py]: here we import only the functions that are strictly needed as the line below

File: minisom.py [hunk 2]
```diff
@@ -390,15 +391,23 @@ def activation_response(self, data):
...
+    
+    def dist_mat(self, data):
+        """
+        calculate distance matrix: dist[i,j]: i: the i-th data, j: the j-th node
+        """
+        D = np.array(data)
+        C = self._weights.reshape(-1, self._weights.shape[2])
+        dd = (D ** 2).sum(axis=1, keepdims=True)
+        cc = (C ** 2).sum(axis=1, keepdims=True)
+        dc = D @ C.T
```
Comment (JustGlowing) [inline minisom.py]: let's use `matmult` rather than `@`

File: minisom.py [hunk 3]
```diff
@@ -416,23 +425,12 @@ def topographic_error(self, data):
...
-            bmu_2 = unravel_index(where(indexes == 1)[0][0],
-                                  self._activation_map.shape)
-            error += are_adjacent(bmu_1, bmu_2)
-        return error / float(len(data))
+        inds = np.argsort(self.dist_mat(data), axis=1)[:, :2]
+        xy = np.unravel_index(inds, self._weights.shape[:2])
+        x, y = xy[0], xy[1]
+        dxdy = np.hstack([np.diff(x), np.diff(y)])
+        diff = np.linalg.norm(dxdy, axis=1)
+        return (diff > 1.42).mean()
```
Comment (JustGlowing) [inline minisom.py]: I redesigned the test for this function:

```
    def test_topographic_error(self):
        # 5 will have bmu_1 in (2,3) and bmu_2 in (2, 4)
        # which are in the same neighborhood
        self.som._weights[2, 4] = 6.0
        # 15 will have bmu_1 in (4, 4) and bmu_2 in (0, 0)
        # which are not in the same neighborhood
        self.som._weights[4, 4] = 15.0
        self.som._weights[0, 0] = 14.
        assert self.som.topographic_error([[5]]) == 0.0
        assert self.som.topographic_error([[15]]) == 1.0
```

Can you make sure it is passed?

File: minisom.py [hunk 4]
```diff
@@ -390,15 +391,23 @@ def activation_response(self, data):
         for x in data:
             a[self.winner(x)] += 1
         return a
+    
+    def dist_mat(self, data):
+        """
+        calculate distance matrix: dist[i,j]: i: the i-th data, j: the j-th node
+        """
+        D = np.array(data)
+        C = self._weights.reshape(-1, self._weights.shape[2])
```
Comment (JustGlowing) [inline minisom.py]: let's use explicit names, for example
`weights_flat = self._weights.reshape(-1, self._weights.shape[2])`

File: minisom.py [hunk 5]
```diff
@@ -390,15 +391,23 @@ def activation_response(self, data):
         for x in data:
             a[self.winner(x)] += 1
         return a
+    
+    def dist_mat(self, data):
```
Comment (JustGlowing) [inline minisom.py]: this function should be called `_distance_from_weights` and should have a unit test.

File: minisom.py [hunk 6]
```diff
@@ -416,23 +425,12 @@ def topographic_error(self, data):
...
-            flat_map = activations.reshape(total_neurons)
-            indexes = argsort(flat_map)
-            bmu_1 = unravel_index(where(indexes == 0)[0][0],
-                                  self._activation_map.shape)
-            bmu_2 = unravel_index(where(indexes == 1)[0][0],
-                                  self._activation_map.shape)
-            error += are_adjacent(bmu_1, bmu_2)
-        return error / float(len(data))
+        inds = np.argsort(self.dist_mat(data), axis=1)[:, :2]
+        xy = np.unravel_index(inds, self._weights.shape[:2])
```
Comment (JustGlowing) [inline minisom.py]: try to use more explicit variable names, `bmu` is better than `x` or `y`.

**Comment by JustGlowing** [comment]
hi @wei-zhang-thz , I left my code review as promised. Let me know once you manage to have a look.

**Comment by JustGlowing** [comment]
Also, if you merge the new version of master in your code you'll find the test mentioned in a commend above.

**Comment by wei-zhang-thz** [comment]
Thanks for the review! I will resolve the problems in one or two days.

File: minisom.py [hunk 7]
```diff
@@ -390,15 +391,23 @@ def activation_response(self, data):
...
+    def dist_mat(self, data):
+        """
+        calculate distance matrix: dist[i,j]: i: the i-th data, j: the j-th node
+        """
+        D = np.array(data)
+        C = self._weights.reshape(-1, self._weights.shape[2])
+        dd = (D ** 2).sum(axis=1, keepdims=True)
+        cc = (C ** 2).sum(axis=1, keepdims=True)
+        dc = D @ C.T
+        return np.sqrt(-2 * dc + dd + cc.T)
```
Comment (JustGlowing) [inline minisom.py]: wouldn't it be simpler to compute the distances as `(np.sum((data[None,:] - self._weights[:, None])**2, -1)**0.5)`?

**Comment by wei-zhang-thz** [comment]
Please check the new version.

Your suggestion on using an alternative way to implement _distance_from_weights() is indeed very simple (but has some problem). A correct implementation would be something like: _norm(input_data[:, :, None] - weights_flat.T[None, :, :], axis=1)_. However, this implementation does not work well for big data because of memory consumption.

Assume m is number of data entries, n is number of nodes, and k is data dimension, then the runtime analysis of the two implementations of _distance_from_weights() are:
Time:      O[mnk] for both versions
Space:    my version: O[mk + nk + mn]    your version: O[mnk]
You can see the big difference in the space complexity.

File: minisom.py [hunk 8]
```diff
@@ -546,6 +547,16 @@ def test_activation_reponse(self):
     def test_activate(self):
         assert self.som.activate(5.0).argmin() == 13.0  # unravel(13) = (2,3)
 
+    def test_distance_from_weights(self):
+        self.som._weights = arange(2 * 3 * 4).reshape(2, 3, 4)
```
Comment (JustGlowing) [inline minisom.py]: do not overwrite the a variable filled in the setup. Also, `self.som` is supposed to have weights with shape `(5, 5, 1)`.

File: minisom.py [hunk 9]
```diff
@@ -397,15 +398,23 @@ def activation_response(self, data):
         for x in data:
             a[self.winner(x)] += 1
         return a
+    
+    def _distance_from_weights(self, data):
```
Comment (JustGlowing) [inline minisom.py]: great job on the variables naming 👍

File: minisom.py [hunk 10]
```diff
@@ -546,6 +547,16 @@ def test_activation_reponse(self):
...
         assert self.som.activate(5.0).argmin() == 13.0  # unravel(13) = (2,3)
 
+    def test_distance_from_weights(self):
+        self.som._weights = arange(2 * 3 * 4).reshape(2, 3, 4)
+        data = arange(2 * 4).reshape(2, 4)
+        assert (self.som._distance_from_weights(data) == array([[0., 8., 16., 24., 32., 40.], [8., 0., 8., 16., 24., 32.]])).all()
+
+    def test_distance_from_weights_big_data(self):
+        self.som._weights = random.rand(50, 30, 300)
+        data = random.rand(10000, 300)
```
Comment (JustGlowing) [inline minisom.py]: Tests need to be minimalistic and repeatable. It's more valuable to test edge cases rather than the cases in which we're sure about the behaviour.

In the method `setUp` there is a some already defined. Please use that.

**Comment by JustGlowing** [comment]
pycodestyle gives flags some lines now:

```
$ pycodestyle minisom.py                                                  
minisom.py:401:1: W293 blank line contains whitespace
minisom.py:404:80: E501 line too long (80 > 79 characters)
minisom.py:553:80: E501 line too long (130 > 79 characters)
minisom.py:624:19: E702 multiple statements on one line (semicolon)
```

File: minisom.py [hunk 11]
```diff
@@ -3,8 +3,9 @@
 from numpy import (array, unravel_index, nditer, linalg, random, subtract,
                    power, exp, pi, zeros, arange, outer, meshgrid, dot,
                    logical_and, mean, std, cov, argsort, linspace, transpose,
-                   einsum, prod, nan)
+                   einsum, prod, where, nan, sqrt, hstack, diff)
```
Comment (JustGlowing) [inline minisom.py]: `where` is now unused.

File: minisom.py (hunk 10 above)
Comment (wei-zhang-thz) [reply minisom.py]: The purpose of this test is to see the performance under big data. Do you want me to just remove this test?

**Comment by wei-zhang-thz** [comment]
Please check the new version. Thanks!

File: minisom.py [hunk 12]
```diff
@@ -546,6 +548,14 @@ def test_activation_reponse(self):
     def test_activate(self):
         assert self.som.activate(5.0).argmin() == 13.0  # unravel(13) = (2,3)
 
+    def test_distance_from_weights(self):
+        data = arange(-5, 5).reshape(-1, 1)
+        weights = self.som._weights.reshape(-1, self.som._weights.shape[2])
+        distances = self.som._distance_from_weights(data)
+        for i in range(len(data)):
+            for j in range(len(weights)):
+                assert(distances[i][j] == norm(data[i] - weights[j]))
```
Comment (JustGlowing) [inline minisom.py]: this is a great test 👍

**Comment by JustGlowing** [comment]
I'm going to merge this now. 🎉  Thanks for your amazing contribution. You'll be mentioned in the notes of the next release and on a twitter announcement. Let me know if you have a twitter account.

**Comment by wei-zhang-thz** [comment]
Thank you! It's a great experience and I learned a lot from you. @835Aloha is my twitter account.

**Comment by JustGlowing** [comment]
hi there, I'm thinking about changing the license of Minisom to MIT (or GPL). This will allow Minisom to have a paper on the Journal of Open Source software. This WILL NOT affect the ownership of your contribution and does not imply that I will profit from Minisom. Your contribution was very welcome and I'd like to have your approval.

**Comment by wei-zhang-thz** [comment]
Hello, Thanks for letting me know this. I have no problem with the license
changing.
